src/biokbase/narrative/app_util.py

```diff
@@ -738,68 +742,153 @@ def transform_param_value(transform_type, value, spec_param):
+        return transform_object_value(transform_type, value)
 
-    elif transform_type == "int":
+    if transform_type == "int":
         # make it an integer, OR 0.
         if value is None or len(str(value).strip()) == 0:
             return None
         return int(value)
 
-    elif transform_type == "string":
+    if transform_type == "string":
         if value is None:
             return value
-        elif isinstance(value, list):
+        if isinstance(value, list):
             return ",".join(value)
-        elif isinstance(value, dict):
-            return ",".join([key + "=" + str(value[key]) for key in value])
-        else:
-            return str(value)
+        if isinstance(value, dict):
+            return ",".join([f"{key}={value[key]}" for key in value])
+        return str(value)
 
-    elif transform_type.startswith("list<") and transform_type.endswith(">"):
+    if transform_type.startswith("list<") and transform_type.endswith(">"):
         # make it a list of transformed types.
         list_type = transform_type[5:-1]
         if isinstance(value, list):
-            ret = []
-            for pos in range(0, len(value)):
-                ret.append(transform_param_value(list_type, value[pos], None))
-            return ret
-        else:
-            return [transform_param_value(list_type, value, None)]
+            return [transform_param_value(list_type, v, None) for v in value]
+        return [transform_param_value(list_type, value, None)]
 
     else:
         raise ValueError("Unsupported Transformation type: " + transform_type)
+
+
+def transform_object_value(transform_type: Optional[str], value: Optional[str]) -> str:
+    """
+    Cases:
+    transform = ref or unresolved-ref:
+        - should return wsname / object name
+    transform = upa or resolved-ref:
+        - should return UPA
+    transform = None:
+        - should return object name
+    Note that if it is a reference path, it was always get returned as an UPA-path
+    for the best compatibility.
+
+    value can be either object name, ref, upa, or ref-path
+    can tell by testing with UPA api
+
+    If we can't find any object info on the value, just return the value as-is
+    """
+    if value is None:
+        return None
+
+    # 1. get object info
+    is_upa = upa.is_upa(value)
+    is_ref = upa.is_ref(value)
+    is_path = (is_upa or is_ref) and ";" in value
+
+    search_ref = value
+    if not is_upa and not is_ref:
+        search_ref = f"{system_variable('workspace')}/{value}"
+    try:
+        obj_info = clients.get("workspace").get_object_info3(
+            {"objects": [{"ref": search_ref}]}
+        )
+    except Exception as err:
+        print(err)
+        print(f"Error while searching for object ref '{search_ref}'")
+        return value
+
+    if is_path:
+        return ";".join(obj_info["paths"][0])
+    if transform_type == "ref" or transform_type == "unresolved-ref":
```

Comment:
I don't know if it'd be marginally more readable as

```py
if transform_type in ["ref", "unresolved-ref"]:
```

Probably not enough to make it worth changing.